src/signalrclient/hub_connection_impl.cpp

```diff
@@ -171,8 +174,21 @@ namespace signalr
 
     void hub_connection_impl::stop(std::function<void(std::exception_ptr)> callback) noexcept
     {
-        m_callback_manager.clear(signalr::value(std::map<std::string, signalr::value> { { std::string("error"), std::string("connection was stopped before invocation result was received") } }));
-        m_connection->stop(callback);
+        if (get_connection_state() == connection_state::disconnected)
+        {
+            m_logger.log(trace_level::info, "Stop ignored because the connection is already disconnected.");
+            callback(nullptr);
+            return;
+        }
+        else if (get_connection_state() == connection_state::disconnecting)
+        {
+            m_logger.log(trace_level::info, "Stop ignored because the connection is in the disconnecting state.");
+            callback(nullptr);
```

Comment:
Can we wait to call the callback once the connection is in the disconnected state?

Reply:
I thought about it, I saw the TS client doesn't wait and the C# one does.

If we wait, would it be a synchronous wait or async?

Reply:
I would make the wait async since the stop itself is async.

[The TS client does wait during the disconnecting state](https://github.com/dotnet/aspnetcore/blob/44e44931fe4050e82b2877047b5104d779147f78/src/SignalR/clients/ts/signalr/src/HubConnection.ts#L259-L262) like the C# client does. The log is a little misleading in saying the call to stop() is "ignored" because the HubConnection is in the disconnecting state. The call to stop() doesn't cause anything new to happen in this case, but it does return the promise from the last call to stop().